File: project_euler/Problem058/Problem058.c
#include <stdio.h>
#include <math.h>

static int is_prime(int value);

void Problem058(void) {
    double target = 0.1;
    int index = 3;
    int count = 0;
    while (1) {
        int value = index * index;
        for (int i = 0; i < 3; ++i) {
            value -= index - 1;
            if (is_prime(value)) {
                ++count;
            }
        }

        if ((double) count < ((index << 1) - 1) * target) {
            printf("%d\n", index);
            return;
        }
        index += 2;
    }
}

static int is_prime(int value) {
    if (1 != value % 6 && 5 != value % 6) {
        return (2 == value || 3 == value) ? 1 : 0;
    }

    int border = (int) sqrt(value);
    for (int i = 5; i <= border; i += 6) {
        if (!(value % i) || !(value % (i + 2))) {
            return 0;
        }
    }
    return (1 != value) ? 1 : 0;
}

File: project_euler/Problem058/Problem058.py
import math

class Problem058:
    def solution(self) -> None:
        target, index, count = 0.1, 3, 0
        while True:
            value = index * index
            for _ in range(3):
                value -= index - 1
                if self.__is_prime(value):
                    count += 1

            if count < ((index << 1) - 1) * target:
                print(index)
                return
            index += 2

    def __is_prime(self, value: int) -> bool:
        if 1 != value % 6 and 5 != value % 6:
            return 2 == value or 3 == value

        border = int(math.sqrt(value))
        for i in range(5, border + 1, 6):
            if not value % i or not value % (i + 2):
                return False
        return 1 != value
